rep-xxxx.rst

```diff
@@ -0,0 +1,268 @@
+It is expected that in most cases ``tool0`` will be a child of the last physical link of a robot's kinematic chain (ie: the 6th or 7th link for a standard industrial serial manipulator).
+
+``tool0`` must not be changed.
+If application-specific tool frames are needed, these should be added as siblings of ``tool0``, or as part of a(n) (stand-alone) EEF definition attached to the ``flange`` frame.
+
+``tool0`` shall not have any geometry associated with it.
+
+Rationale: 
+
+By not allowing changes to the location and orientation of ``tool0``, re-use of libraries such as kinematics solvers that are generated in an off-line fashion for a particular kinematic chain configuration becomes feasible.
+It is the user's responsibility then to make sure that poses are transformed into the appropriate coordinate system before passing them on to such libraries.
+This could be done automatically by the motion planner, or manually before submitting goal poses to the planner.
+
+
+Dual or Multi-arm robots
+------------------------
+
+This REP does not specify any special conventions for robots with multiple arms/groups or kinematic chains.
+The standard approach of prefixing joint and link names to ensure uniqueness of all frames in the (combined) frame hierarchy should be used to avoid collisions.
+See the `Example Frame Hierarchies`_ section for examples of this.
+
+
+Frame Authorities
+-----------------
+
+The frames described in this REP will typically be part of the static description of robot models encoded in urdfs or xacros.
+As such, the frame authority is expected to be an instance of ``robot_state_publisher``, but this is not required.
+In cases where (complicated) kinematics preclude the use of standard nodes, a specialised node capable of publishing the necessary frames could be used.
+
+
+Exceptions
+----------
+
+The scope of potential robotics software is too broad to require all ROS software to follow the guidelines of this REP.
+However, choosing different conventions should be well justified, well documented, and is discouraged.
+
+
+Example Frame Hierarchies
+=========================
+
+Single manipulator
+------------------
+
+The following shows an example frame hierarchy for a single serial manipulator with just the default tool frame::
+
+  base_link
+  ├ base
+  └ link_1
+    └ link_2
+      └..
+       └ link_N
+         ├ tool0
+         └ flange
+
+Single manipulator with EEF
+---------------------------
+
+The following shows an example frame hierarchy for a single serial manipulator with an EEF model attached to ``flange`` and an additional tool frame::
+
+  base_link
+  ├ base
+  └ link_1
+    └ link_2
+      └ ..
+        └ link_N
+          ├ tool0
+          ├ flange
+          │ └ eef_base_link
+          │   └ ..
+          │     └ eef_link_N
+          └ eef_tcp
```

Comment:
I'm not sure here whether any `tcp` frame should be a child of `flange` or not.

Many methods of 'teaching' tool frames will (in the end) be relative to the flange of the robot, so making all tool frames (including `tool0`) children of `flange` could make sense.

However that would also imply that the origin of `flange` must be exactly what it is on the real robot, which may not always be known or documented, and makes the frame not only a convenience (ie: an attachment point for EEF xacro models), but promotes it to an entity that (should accurately) model(s) a physical part of the real robot.

Reply:
(slightly out-of-scope here) I didn't make `eef_tcp` a child of `eef_base_link` as that is not how I believe most (all?) ind robots do this. Tool frames will always be relative to the flange, and by doing it this way, we also keep the transform `base -> eef_tcp` relatively uncomplicated.

Reply:
I think `link_N` should be mostly internal, and that all external frames should reference either `flange` or `tool0`.  Under that logic, `eef_tcp` is wrongly placed here.  Some more discussion is required about how TCP is handled, though.

  - There is some advantage to deriving `eef_tcp` from `eef_base_link`.  This allows more modularity in packaging of eef URDFs, so that a new "tool" can be attached with a single joint, defining both geometry and TCP.

  - Should TCP be attached to `flange` or `tool0`?  This may depend on how closely your ROS application is coupled to the controller.  It might be handy to do planning relative to `base` and `tool0` if passing Cartesian coords to robot, but ROS-centric planning might prefer to use `base_link` and `flange`.
  - Can we at least provide recommendations on how to assign TCP?  As a formal section, not just a footnote below this diagram.

Reply:
@JeremyZoss wrote:
> I think `link_N` should be mostly internal, and that all external frames should reference either `flange` or `tool0`.  Under that logic, `eef_tcp` is wrongly placed here.  Some more discussion is required about how TCP is handled, though.
> 
>   - There is some advantage to deriving `eef_tcp` from `eef_base_link`.  This allows more modularity in packaging of eef URDFs, so that a new "tool" can be attached with a single joint, defining both geometry and TCP.

yes, but it would/could also complicate setting up xacros that can be calibrated. If we include tool frames in EEF models, the files will have to be edited (difficult if the EEF was distributed as a binary package (ie: non world-writable)) if the toolframe needs to be adjusted for some reason. If we keep tool frames out of the EEF xacro, users that want to reuse them can easily do calibration of the composition `robot+eef` and then add their new tool frame by attaching it to `robot_flange` in the composing xacro.

We could of course standardise naming of frames in EEF xacros, and provide something similar to `flange` (ie: an attachment point) in those xacros, but I expect more variability in EEFs than in robots, so that could be a bit much (for now at least).

Reply:
@JeremyZoss wrote:
>  - Should TCP be attached to `flange` or `tool0`?  This may depend on how closely your ROS application is coupled to the controller.  It might be handy to do planning relative to `base` and `tool0` if passing Cartesian coords to robot, but ROS-centric planning might prefer to use `base_link` and `flange`.

I think it should be `flange` in any case, as `tool0` does not follow REP 103 (they typically have z+ forward) which complicates attaching additional links in a xacro/urdf (you need to introduce additional rotations to undo the x+->z+ change in `tool0`).

As to planning: I feel either:

 - planning libraries should be able to transform incoming poses to whatever the parent is of the configured EEF (MoveIt is a bit inconsistent with it, but it can do it, that's why EEF groups are configured in the MSA)
 - users should transform poses back to `tool0` (or whatever we decide is going to be the 'end link' (but see my reply to your later comment))

Reply:
@JeremyZoss wrote:
>   - Can we at least provide recommendations on how to assign TCP?  As a formal section, not just a footnote below this diagram.

Can you clarify what you mean? I'm not sure I understand the comment.

Reply:
> > Can we at least provide recommendations on how to assign TCP? As a formal section, not just a footnote below this diagram.

> Can you clarify what you mean? I'm not sure I understand the comment.

I feel like most "real" applications (as opposed to simplified exercises) will try to control a robot's TCP.  It's not obvious/trivial what is the "typical" approach for creating such a TCP frame: naming convention, parent frame, etc, as evidenced by some of the discussions on this page.  My comment here means that it might be worth adding a section to discuss expectations related to the TCP frame.  It's not really mentioned in this writeup, except as listed in the "Single Arm Manipulator with EEF" diagram.  The diagram should (ideally) illustrate concepts discussed elsewhere in the writeup.

Reply:
I'm not against adding something about TCP frames, but I'm not sure it's in the scope of this document per se. As with REP-105 (frames for mobile bases) and REP-120 (frames for humanoids) the intend is to define a nr of standardised frames and their semantics - but for industrial robots.

I think there is definitely enough to say about it, but perhaps in a separate REP?